config: add docker-compose, configure flyway,

Nique1 opened the pull request:
Add docker-compose file, use postgres image, create docker container to run postgres db.

Configure flyway, add flyway scripts, create songs table, insert data to songs table.

Nique1 committed:
config: add docker-compose, configure flyway,
Add docker-compose file, use postgres image, create docker container to run postgres db.

Configure flyway, add flyway scripts, create songs table, insert data to songs table.

genobis commented on docker-compose.yml:
```diff
@@ -0,0 +1,10 @@
+services:
+  db:
+    image: postgres:latest
```
Bezpieczniej będzie użyć jakiejś konkretnej wersji, choćby tylko podając główny nr np `postgres:16` 🙂

genobis commented on src/main/resources/application.properties:
```diff
@@ -1 +1,20 @@
+#DB
+spring.jpa.database-platform=org.hibernate.dialect.PostgreSQLDialect
+spring.datasource.driver-class-name=org.postgresql.Driver
+spring.datasource.url=jdbc:postgresql://localhost:5433/music-db
+spring.datasource.username=postgres
+spring.datasource.password=secret
```
Proponuję dodać (gdzieś w początkach) property `spring.profiles.active=local`, co spowoduje aktywowanie profilu `local`.

I te trzy właściwości związane z datasource (url, username, password) przenieść do pliku `application-local.properties`

Tym sposobem ogólna konfiguracja aplikacyjna nie będzie w sobie zawierała właściwości potrzebnych tylko do lokalnego uruchomienia.

To co tu napisałem wystarczy do osiągnięcia zamierzonego efektu, ale jeśli jesteś zainteresowana, więcej o springowych profilach możesz poczytać np. tu: https://www.baeldung.com/spring-profiles

genobis commented on src/main/resources/application.properties:
```diff
@@ -1 +1,20 @@
+spring.datasource.username=postgres
+spring.datasource.password=secret
+#JPA
+#spring.jpa.generate-ddl=false
```
To można śmiało usunąć 🙂

genobis commented on src/main/resources/application.properties:
```diff
@@ -1 +1,20 @@
+spring.datasource.password=secret
+#JPA
+#spring.jpa.generate-ddl=false
+spring.jpa.hibernate.ddl-auto=none
```
Zobacz co robi opcja `validate` 🙂

genobis commented on src/main/resources/db/migration/V1.0.0__create_songs_table.sql:
```diff
@@ -0,0 +1,9 @@
+ id INTEGER NOT NULL GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
+ artistName VARCHAR(255) NOT NULL,
+ trackName VARCHAR(255) NOT NULL,
+ trackTime TIME NOT NULL,
```
To typ danych reprezentujących czas, tj. godziny i minuty w zakresie 00:00-23:59 w oderwaniu od daty - o ile oczywiście dałoby się tego użyć w założony przez Ciebie sposób, to jednak wydaje mi się, że nie do końca tu pasuje.

Nazwałbym to chyba `trackLength` i do reprezentacji użył albo typu `interval` (mapowane na `Duration` - ale tu prawdopodobnie [trzeba trochę Hibernate pomóc](https://vladmihalcea.com/map-postgresql-interval-java-duration-hibernate/)) lub po prostu `int`/`bigint` mapowanych odpowiednio na `int`/`long` (jeśli to czas w sekundach to `int`  zupełności wystarczy, jeśli w milisekundach to konwencja raczej zaleca stosowanie `long`).

Zgodnie z zasadą KISS osobiście pewnie wybrałbym `int` reprezentujący po prostu długość utworu w sekundach, ale ostateczna decyzja należy do Ciebie 😉

genobis commented on src/main/resources/db/migration/V1.0.0__create_songs_table.sql:
```diff
@@ -0,0 +1,9 @@
+ artistName VARCHAR(255) NOT NULL,
+ trackName VARCHAR(255) NOT NULL,
+ trackTime TIME NOT NULL,
+ genre VARCHAR (255) NOT NULL,
```
Czy chcemy przechowywać gatunek jako tekstową wartość razem z utworem? Czy zawsze mamy pewność, że piosenka należy do jednego i tylko jednego gatunku? Dobra praktyka raczej kazałaby mieć listę gatunków gdzieś osobno.

Myślę, że to jest coś nad czym warto się zastanowić i to dość poważnie, ale niekoniecznie od razu teraz 😉 spokojnie można dodać kolumnę później.

genobis commented on src/main/resources/db/migration/V1.0.0__create_songs_table.sql:
```diff
@@ -0,0 +1,9 @@
+ trackName VARCHAR(255) NOT NULL,
+ trackTime TIME NOT NULL,
+ genre VARCHAR (255) NOT NULL,
+ decade INTEGER NOT NULL,
```
A czemu nie rok 🙂 ? Mając lata bez problemu pogrupujemy utwory wg dekad, w drugą stronę to się nie uda.

genobis commented on src/main/resources/db/migration/V1.0.0__create_songs_table.sql:
```diff
@@ -0,0 +1,9 @@
+ trackTime TIME NOT NULL,
+ genre VARCHAR (255) NOT NULL,
+ decade INTEGER NOT NULL,
+ emotion VARCHAR(255) NOT NULL
```
Myślę, że tu trochę podobnie jak z gatunkami - emocje też powinny być jakimś osobnym zamkniętym zbiorem i raczej trzeba założyć, że to będzie coś charakterystycznego dla konkretnego użytkownika (zatem emocja byłaby cechą innej encji, na styku użytkownika i piosenki).

Podobnie jak gatunki, emocje sugerowałbym dodać później.

genobis commented on src/main/resources/db/migration/V1.0.1__insert_songs_data.sql:
```diff
@@ -0,0 +1,6 @@
+INSERT INTO songs(artistname, trackname, tracktime, genre, decade, emotion) VALUES
```
Na razie tak może być, natomiast co do zasady dodawanie danych testowych nie powinno być elementem migracji.

Nique1 committed:
config: add docker-compose, configure flyway,
Add docker-compose file, use postgres image, create docker container to run postgres db.

Configure flyway, add flyway scripts, create songs table, insert data to songs table.

Refactoring and making corrections with configuration
